## **Exercise 1 Solution**

# Exercise 1(1)

For solving this problem, we can associate a decision variable $x_{ij}$ for denoting the number of truck load of EVs transported from warehouses $i$ to markets $j$.

Let  i  take values from the set  {1,2,3,4,5,6,7,8}  corresponding to Ahmedabad, Bangaluru, Channi, Delhi, Kolkata, Lucknow, Mumbai and Hyderabad. Let  j  take values from the set  {1,2,3,4,5,6,7,8}  corresponding to Hassan, Jodhpur, Trichy, Simla, Vijayawada, Dibrugarh, Raipur and Agartala.

Then we can see that the number of truck load of EVs transported from Ahmedhabad to the eight markets is given by the sum $\sum_{j=1}^{8} x_{1j}$. 
Similarly, the number of truck load of EVs transported from Bengaluru to the eight markets is given by $\sum_{j=1}^{8} x_{2j}$ ,
the number of truck load of EVs transported from Chennai to the eight markets is given by $\sum_{j=1}^{8} x_{3j}$ , 
the number of truck load of EVs transported from Delhi to the eight markets is given by $\sum_{j=1}^{8} x_{4j}$ ,
the number of truck load of EVs transported from Kolkata to the eight markets is given by $\sum_{j=1}^{8} x_{5j}$ ,
the number of truck load of EVs transported from Lucknow to the eight markets is given by $\sum_{j=1}^{8} x_{6j}$ ,
the number of truck load of EVs transported from Mumbai to the eight markets is given by $\sum_{j=1}^{8} x_{7j}$ 
And the number of truck load of EVs transported from Hyderabad to the eight markets is given by $\sum_{j=1}^{8} x_{8j}$.
From the constraints on the Available Stock, we can thus write: 
\begin{align}
\sum_{j=1}^{8} x_{1j} &\leq 350 \nonumber \\
\sum_{j=1}^{8} x_{2j} &\leq 185 \nonumber \\
\sum_{j=1}^{8} x_{3j} &\leq 305 \nonumber \\
\sum_{j=1}^{8} x_{4j} &\leq 180 \nonumber \\
\sum_{j=1}^{8} x_{5j} &\leq 205 \nonumber \\
\sum_{j=1}^{8} x_{6j} &\leq 175 \nonumber \\
\sum_{j=1}^{8} x_{7j} &\leq 325 \nonumber \\
\sum_{j=1}^{8} x_{8j} &\leq 370 \nonumber \\
\end{align}

Also, note that the number of truck load of EVs reaching Hassan from the eight warehouses is given by the sum $\sum_{i=1}^{8} x_{i1}$.
The number of truck load of EVs reaching Jodhpur from the eight warehouses is given by the sum $\sum_{i=1}^{8} x_{i2}$.
The number of truck load of EVs reaching Jodhpur from the eight warehouses is given by the sum $\sum_{i=1}^{8} x_{i3}$.
The number of truck load of EVs reaching Jodhpur from the eight warehouses is given by the sum $\sum_{i=1}^{8} x_{i4}$.
The number of truck load of EVs reaching Jodhpur from the eight warehouses is given by the sum $\sum_{i=1}^{8} x_{i5}$.
The number of truck load of EVs reaching Jodhpur from the eight warehouses is given by the sum $\sum_{i=1}^{8} x_{i6}$.
The number of truck load of EVs reaching Jodhpur from the eight warehouses is given by the sum $\sum_{i=1}^{8} x_{i7}$.
The number of truck load of EVs reaching Jodhpur from the eight warehouses is given by the sum $\sum_{i=1}^{8} x_{i8}$.

To satisfy the demands at the eight markets, we thus need the following constraints:

\begin{align}
\sum_{i=1}^{8} x_{i1} &= 205 \nonumber \\
\sum_{i=1}^{8} x_{i2} &= 300 \nonumber \\
\sum_{i=1}^{8} x_{i3} &= 425 \nonumber \\
\sum_{i=1}^{8} x_{i4} &= 270 \nonumber \\
\sum_{i=1}^{8} x_{i5} &= 375 \nonumber \\
\sum_{i=1}^{8} x_{i6} &= 225 \nonumber \\
\sum_{i=1}^{8} x_{i7} &= 185 \nonumber \\
\sum_{i=1}^{8} x_{i8} &= 110 \nonumber \\
\end{align}

Also note that the number truck load of Es transported between a warehouses $i$ and a markets $j$ must be non-negative. This gives us the constraint $x_{ij} \geq 0, \ \forall i \in \{1,2,3,4,5,6,7,8\}, \ \forall j \in \{1,2,3,4,6,7,8\}$.

Now we shall formulate the objective function. Since $x_{ij}$ trucks are transported between warehouse $i$ and markets $j$, and since we know the transportation cost of each truck along the route between warehouse $i$ and markets $j$, we have the total transportation cost as below: 

$827 x_{11} + 817 x_{12} + 1470 x_{13} + 2272 x_{14} + 1615 x_{15} + 3143 x_{16} + 1264 x_{17} + 4465 x_{18} + 559 x_{21} + 1832 x_{22} + 751 x_{23} + 3772 x_{24} + 637 x_{25} + 3773 x_{26} + 1856 x_{27} + 5459 x_{28} + 1109 x_{31} + 2443 x_{32} + 577 x_{33} + 3865 x_{34} + 798 x_{35} + 3521 x_{36} + 1984 x_{37} + 5303 x_{38} + 2943 x_{41} + 1471 x_{42} + 4676 x_{43} + 1337 x_{44} + 2702 x_{45} + 2741 x_{46} + 2482 x_{47} + 4148 x_{48} + 3742 x_{51} + 3089 x_{52} + 2931 x_{53} + 3199 x_{54} + 2226 x_{55} + 1739 x_{56} + 1691 x_{57} + 2128 x_{58} + 2353 x_{61} + 1160 x_{62} + 3295 x_{63} + 368 x_{64} + 2369 x_{65} + 2365 x_{66} + 1293 x_{67} + 3642 x_{68} + 1356 x_{71} + 1617 x_{72} + 2045 x_{73} + 2941 x_{74} + 2334 x_{75} + 4151 x_{76} + 1937 x_{77} + 4287 x_{78} + 195 x_{81} + 1862 x_{82} + 1331 x_{83} + 2921 x_{84} + 402 x_{85} + 2825 x_{86} + 563 x_{87} + 4441 x_{88}$.

***Thus we have the following optimization problem:***

\begin{align}
\min  \ 827 x_{11} + 817 x_{12} + 1470 x_{13} + 2272 x_{14} + 1615 x_{15} + 3143 x_{16} + 1264 x_{17} + 4465 x_{18} + 559 x_{21} + 1832 x_{22} + 751 x_{23} + 3772 x_{24} + 637 x_{25} + 3773 x_{26} + 1856 x_{27} + 5459 x_{28} + 1109 x_{31} + 2443 x_{32} + 577 x_{33} + 3865 x_{34} + 798 x_{35} + 3521 x_{36} + 1984 x_{37} + 5303 x_{38} + 2943 x_{41} + 1471 x_{42} + 4676 x_{43} + 1337 x_{44} + 2702 x_{45} + 2741 x_{46} + 2482 x_{47} + 4148 x_{48} + 3742 x_{51} + 3089 x_{52} + 2931 x_{53} + 3199 x_{54} + 2226 x_{55} + 1739 x_{56} + 1691 x_{57} + 2128 x_{58} + 2353 x_{61} + 1160 x_{62} + 3295 x_{63} + 368 x_{64} + 2369 x_{65} + 2365 x_{66} + 1293 x_{67} + 3642 x_{68} + 1356 x_{71} + 1617 x_{72} + 2045 x_{73} + 2941 x_{74} + 2334 x_{75} + 4151 x_{76} + 1937 x_{77} + 4287 x_{78} + 195 x_{81} + 1862 x_{82} + 1331 x_{83} + 2921 x_{84} + 402 x_{85} + 2825 x_{86} + 563 x_{87} + 4441 x_{88} \nonumber \\
{\rm{s.t. }} \ \sum_{j=1}^{8} x_{1j} &\leq 350 \nonumber \\
\sum_{j=1}^{8} x_{2j} &\leq 185 \nonumber \\
\sum_{j=1}^{8} x_{3j} &\leq 305 \nonumber \\
\sum_{j=1}^{8} x_{4j} &\leq 180 \nonumber \\
\sum_{j=1}^{8} x_{5j} &\leq 205 \nonumber \\
\sum_{j=1}^{8} x_{6j} &\leq 175 \nonumber \\
\sum_{j=1}^{8} x_{7j} &\leq 325 \nonumber \\
\sum_{j=1}^{8} x_{8j} &\leq 370 \nonumber \\
\text{ and } \nonumber \\  
\sum_{i=1}^{8} x_{i1} &= 205 \nonumber \\
\sum_{i=1}^{8} x_{i2} &= 300 \nonumber \\
\sum_{i=1}^{8} x_{i3} &= 425 \nonumber \\
\sum_{i=1}^{8} x_{i4} &= 270 \nonumber \\
\sum_{i=1}^{8} x_{i5} &= 375 \nonumber \\
\sum_{i=1}^{8} x_{i6} &= 225 \nonumber \\
\sum_{i=1}^{8} x_{i7} &= 185 \nonumber \\
\sum_{i=1}^{8} x_{i8} &= 110 \nonumber \\
\text{ and } \nonumber \\ 
\ x_{ij} \geq 0, \ \forall i \in \{1,2,3,4,5,6,7,8\}, \ \forall j \in \{1,2,3,4,6,7,8\} \nonumber \\
\end{align}


In [ ]:
!pip install -q pyomo
from pyomo.environ import * 
import numpy as np

     |████████████████████████████████| 9.7 MB 3.7 MB/s 
     |████████████████████████████████| 49 kB 4.5 MB/s 


In [ ]:
import pandas as pd

In [ ]:
data_csvfile = pd.read_csv('lab6_ex1.csv')

In [ ]:
print(data_csvfile)

    Warehouse  Hassan  Jodhpur  Trichy  Simla  VIjayawada  Dibrugarh  Raipur  \
0  Ahmedabad      827      817    1470   2272        1615       3143    1264   
1   Bengaluru     559     1832     751   3772         637       3773    1856   
2     Channai    1109     2443     577   3865         798       3521    1984   
3       Delhi    2943     1471    4676   1337        2702       2741    2482   
4     Kolkata    3742     3089    2931   3199        2226       1739    1691   
5     Lucknow    2353     1160    3295    368        2369       2365    1293   
6      Mumbai    1356     1617    2045   2941        2334       4151    1937   
7   Hyderabad     195     1862    1331   2921         402       2825     563   
8      Demand     205      300     425    270         375        225     185   

   Agartala   Astock  
0      4465      350  
1      5459      185  
2      5303      305  
3      4148      180  
4      2128      205  
5      3642      175  
6      4287      325  
7      4441    

In [ ]:
data_csvfile.index

RangeIndex(start=0, stop=9, step=1)

In [ ]:
print('num rows:', len(data_csvfile.index))

num rows: 9


In [ ]:
data_csvfile.columns

Index(['Warehouse', 'Hassan', 'Jodhpur', 'Trichy', 'Simla', 'VIjayawada',
       'Dibrugarh', 'Raipur', 'Agartala', ' Astock'],
      dtype='object')

In [ ]:
print('num  columns: ',len(data_csvfile.columns))

num  columns:  10


In [ ]:
model_lab5_ex1 = ConcreteModel()

In [ ]:
M = len(data_csvfile.index)-1
print('num warehouses: ',M)

num warehouses:  8


In [ ]:
N = len(data_csvfile.columns)-2
print('num Markets: ',N)

num Markets:  8


In [ ]:
row_indices = range(M)
col_indices = range(N)

print(row_indices)
print(col_indices)

range(0, 8)
range(0, 8)


In [ ]:
for row_idx in row_indices:
  for col_idx in col_indices:
    print('cost[',row_idx,'][',col_idx+1,']:',data_csvfile.iloc[row_idx, col_idx+1])

cost[ 0 ][ 1 ]: 827
cost[ 0 ][ 2 ]: 817
cost[ 0 ][ 3 ]: 1470
cost[ 0 ][ 4 ]: 2272
cost[ 0 ][ 5 ]: 1615
cost[ 0 ][ 6 ]: 3143
cost[ 0 ][ 7 ]: 1264
cost[ 0 ][ 8 ]: 4465
cost[ 1 ][ 1 ]: 559
cost[ 1 ][ 2 ]: 1832
cost[ 1 ][ 3 ]: 751
cost[ 1 ][ 4 ]: 3772
cost[ 1 ][ 5 ]: 637
cost[ 1 ][ 6 ]: 3773
cost[ 1 ][ 7 ]: 1856
cost[ 1 ][ 8 ]: 5459
cost[ 2 ][ 1 ]: 1109
cost[ 2 ][ 2 ]: 2443
cost[ 2 ][ 3 ]: 577
cost[ 2 ][ 4 ]: 3865
cost[ 2 ][ 5 ]: 798
cost[ 2 ][ 6 ]: 3521
cost[ 2 ][ 7 ]: 1984
cost[ 2 ][ 8 ]: 5303
cost[ 3 ][ 1 ]: 2943
cost[ 3 ][ 2 ]: 1471
cost[ 3 ][ 3 ]: 4676
cost[ 3 ][ 4 ]: 1337
cost[ 3 ][ 5 ]: 2702
cost[ 3 ][ 6 ]: 2741
cost[ 3 ][ 7 ]: 2482
cost[ 3 ][ 8 ]: 4148
cost[ 4 ][ 1 ]: 3742
cost[ 4 ][ 2 ]: 3089
cost[ 4 ][ 3 ]: 2931
cost[ 4 ][ 4 ]: 3199
cost[ 4 ][ 5 ]: 2226
cost[ 4 ][ 6 ]: 1739
cost[ 4 ][ 7 ]: 1691
cost[ 4 ][ 8 ]: 2128
cost[ 5 ][ 1 ]: 2353
cost[ 5 ][ 2 ]: 1160
cost[ 5 ][ 3 ]: 3295
cost[ 5 ][ 4 ]: 368
cost[ 5 ][ 5 ]: 2369
cost[ 5 ][ 6 ]: 2365
cost[ 5 ][ 7 ]: 1293
cost[ 5 ][ 8 ]: 3642


In [ ]:
model_lab5_ex1.x = Var(row_indices, col_indices, domain= NonNegativeReals)
model_lab5_ex1.objective = Objective(expr=sum(data_csvfile.iloc[i,j+1]*model_lab5_ex1.x[i,j] for i in row_indices for j in col_indices),sense=minimize)
model_lab5_ex1.constraints = ConstraintList()

In [ ]:
for i in row_indices:
  model_lab5_ex1.constraints.add(expr=sum(model_lab5_ex1.x[i,j] for j in col_indices) <= data_csvfile.iloc[i,N+1]) 

In [ ]:
for j in col_indices:
  model_lab5_ex1.constraints.add(expr=sum(model_lab5_ex1.x[i,j] for i in row_indices) == data_csvfile.iloc[M,j+1])


In [ ]:
model_lab5_ex1.pprint()

4 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   16 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16}
    x_index : Size=1, Index=None, Ordered=True
        Key  : Dimen : Domain              : Size : Members
        None :     2 : x_index_0*x_index_1 :   64 : {(0, 0), (0, 1), (0, 2), (0, 3), (0, 4), (0, 5), (0, 6), (0, 7), (1, 0), (1, 1), (1, 2), (1, 3), (1, 4), (1, 5), (1, 6), (1, 7), (2, 0), (2, 1), (2, 2), (2, 3), (2, 4), (2, 5), (2, 6), (2, 7), (3, 0), (3, 1), (3, 2), (3, 3), (3, 4), (3, 5), (3, 6), (3, 7), (4, 0), (4, 1), (4, 2), (4, 3), (4, 4), (4, 5), (4, 6), (4, 7), (5, 0), (5, 1), (5, 2), (5, 3), (5, 4), (5, 5), (5, 6), (5, 7), (6, 0), (6, 1), (6, 2), (6, 3), (6, 4), (6, 5), (6, 6), (6, 7), (7, 0), (7, 1), (7, 2), (7, 3), (7, 4), (7, 5), (7, 6), (7, 7)}
    x_index_0 : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
    

In [ ]:
!apt-get install -y -qq coinor-cbc

Selecting previously unselected package coinor-libcoinutils3v5.
(Reading database ... 155569 files and directories currently installed.)
Preparing to unpack .../0-coinor-libcoinutils3v5_2.10.14+repack1-1_amd64.deb ...
Unpacking coinor-libcoinutils3v5 (2.10.14+repack1-1) ...
Selecting previously unselected package coinor-libosi1v5.
Preparing to unpack .../1-coinor-libosi1v5_0.107.9+repack1-1_amd64.deb ...
Unpacking coinor-libosi1v5 (0.107.9+repack1-1) ...
Selecting previously unselected package coinor-libclp1.
Preparing to unpack .../2-coinor-libclp1_1.16.11+repack1-1_amd64.deb ...
Unpacking coinor-libclp1 (1.16.11+repack1-1) ...
Selecting previously unselected package coinor-libcgl1.
Preparing to unpack .../3-coinor-libcgl1_0.59.10+repack1-1_amd64.deb ...
Unpacking coinor-libcgl1 (0.59.10+repack1-1) ...
Selecting previously unselected package coinor-libcbc3.
Preparing to unpack .../4-coinor-libcbc3_2.9.9+repack1-1_amd64.deb ...
Unpacking coinor-libcbc3 (2.9.9+repack1-1) ...
Selecting p

In [ ]:
opt_cbc = SolverFactory('cbc')

In [ ]:
result = opt_cbc.solve(model_lab5_ex1)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [ ]:
# display solution
print('\nObjective = ', model_lab5_ex1.objective())

print('\nDecision Variables')
for i in row_indices:
  for j in col_indices:
    print('Num EVs between ',data_csvfile['Warehouse'][i],' and ', data_csvfile.columns[j+1],':', model_lab5_ex1.x[i,j].value)

print('\nConstraints')
model_lab5_ex1.constraints.display()


Objective =  2192870.0

Decision Variables
Num EVs between  Ahmedabad   and  Hassan : 0.0
Num EVs between  Ahmedabad   and  Jodhpur : 300.0
Num EVs between  Ahmedabad   and  Trichy : 0.0
Num EVs between  Ahmedabad   and  Simla : 0.0
Num EVs between  Ahmedabad   and  VIjayawada : 0.0
Num EVs between  Ahmedabad   and  Dibrugarh : 45.0
Num EVs between  Ahmedabad   and  Raipur : 5.0
Num EVs between  Ahmedabad   and  Agartala : 0.0
Num EVs between  Bengaluru  and  Hassan : 0.0
Num EVs between  Bengaluru  and  Jodhpur : 0.0
Num EVs between  Bengaluru  and  Trichy : 120.0
Num EVs between  Bengaluru  and  Simla : 0.0
Num EVs between  Bengaluru  and  VIjayawada : 65.0
Num EVs between  Bengaluru  and  Dibrugarh : 0.0
Num EVs between  Bengaluru  and  Raipur : 0.0
Num EVs between  Bengaluru  and  Agartala : 0.0
Num EVs between  Channai  and  Hassan : 0.0
Num EVs between  Channai  and  Jodhpur : 0.0
Num EVs between  Channai  and  Trichy : 305.0
Num EVs between  Channai  and  Simla : 0.0
Num EVs be

# Exercise 1(7)

Num truck load of EVs between  Ahmedabad   and  Jodhpur : 300

Num truck load of EVs between  Ahmedabad   and  Dibrugarh : 45

Num EVs between  Ahmedabad   and  Raipur : 5

Num EVs between  Bengaluru  and  Trichy : 120

Num EVs between  Bengaluru  and  VIjayawada : 65

Num EVs between  Channai  and  Trichy : 305

Num EVs between  Delhi  and  Simla : 95

Num EVs between  Delhi  and  Dibrugarh : 85

Num EVs between  Kolkata  and  Dibrugarh : 95

Num EVs between  Kolkata  and  Agartala : 110

Num EVs between  Lucknow  and  Simla : 175

Num EVs between  Mumbai  and  Hassan : 205

Num EVs between  Mumbai  and  Raipur : 120

Num EVs between  Hyderabad  and  VIjayawada : 310

Num EVs between  Hyderabad  and  Raipur : 60

Total cost to trasnport EVs by truck werehouse to market is : 2192870

# Exercise 1(8)

the Lucknow-Simla link is disrupted and no transportation is possible on the Lucknow-Simla links. Without changing the pyomo model, this is done by only change in .csv file, in csv file I will very incerese the Lucknow-Simla transportation cost.
i.e new Lucknow-Simla transportation = 999999999

and new objecive function is : 

\begin{align}
\min  \ 827 x_{11} + 817 x_{12} + 1470 x_{13} + 2272 x_{14} + 1615 x_{15} + 3143 x_{16} + 1264 x_{17} + 4465 x_{18} + 559 x_{21} + 1832 x_{22} + 751 x_{23} + 3772 x_{24} + 637 x_{25} + 3773 x_{26} + 1856 x_{27} + 5459 x_{28} + 1109 x_{31} + 2443 x_{32} + 577 x_{33} + 3865 x_{34} + 798 x_{35} + 3521 x_{36} + 1984 x_{37} + 5303 x_{38} + 2943 x_{41} + 1471 x_{42} + 4676 x_{43} + 1337 x_{44} + 2702 x_{45} + 2741 x_{46} + 2482 x_{47} + 4148 x_{48} + 3742 x_{51} + 3089 x_{52} + 2931 x_{53} + 3199 x_{54} + 2226 x_{55} + 1739 x_{56} + 1691 x_{57} + 2128 x_{58} + 2353 x_{61} + 1160 x_{62} + 3295 x_{63} + 999999999 x_{64} + 2369 x_{65} + 2365 x_{66} + 1293 x_{67} + 3642 x_{68} + 1356 x_{71} + 1617 x_{72} + 2045 x_{73} + 2941 x_{74} + 2334 x_{75} + 4151 x_{76} + 1937 x_{77} + 4287 x_{78} + 195 x_{81} + 1862 x_{82} + 1331 x_{83} + 2921 x_{84} + 402 x_{85} + 2825 x_{86} + 563 x_{87} + 4441 x_{88} \nonumber \\
\end{align}

and constraints are same.

In [ ]:
data_csvfile.iloc[5,4]=999999999

In [ ]:
model_lab5_ex1.objective.deactivate()
model_lab5_ex1.objective = Objective(expr=sum(data_csvfile.iloc[i,j+1]*model_lab5_ex1.x[i,j] for i in row_indices for j in col_indices),sense=minimize)


This is usually indicative of a modelling error.
To avoid this warning, use block.del_component() and block.add_component().


In [ ]:
result = opt_cbc.solve(model_lab5_ex1)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [ ]:
# display solution
print('\nObjective = ', model_lab5_ex1.objective())

print('\nDecision Variables')
for i in row_indices:
  for j in col_indices:
    print('Num truck load of EVs between ',data_csvfile['Warehouse'][i],' and ', data_csvfile.columns[j+1],':', model_lab5_ex1.x[i,j].value)

print('\nConstraints')
model_lab5_ex1.constraints.display()


Objective =  2518200.0

Decision Variables
Num truck load of EVs between  Ahmedabad   and  Hassan : 0.0
Num truck load of EVs between  Ahmedabad   and  Jodhpur : 300.0
Num truck load of EVs between  Ahmedabad   and  Trichy : 0.0
Num truck load of EVs between  Ahmedabad   and  Simla : 0.0
Num truck load of EVs between  Ahmedabad   and  VIjayawada : 0.0
Num truck load of EVs between  Ahmedabad   and  Dibrugarh : 50.0
Num truck load of EVs between  Ahmedabad   and  Raipur : 0.0
Num truck load of EVs between  Ahmedabad   and  Agartala : 0.0
Num truck load of EVs between  Bengaluru  and  Hassan : 0.0
Num truck load of EVs between  Bengaluru  and  Jodhpur : 0.0
Num truck load of EVs between  Bengaluru  and  Trichy : 120.0
Num truck load of EVs between  Bengaluru  and  Simla : 0.0
Num truck load of EVs between  Bengaluru  and  VIjayawada : 65.0
Num truck load of EVs between  Bengaluru  and  Dibrugarh : 0.0
Num truck load of EVs between  Bengaluru  and  Raipur : 0.0
Num truck load of EVs betw

# Exercise 1(9)

Num truck load of EVs between  Ahmedabad   and  Jodhpur : 300

Num truck load of EVs between  Ahmedabad   and  Dibrugarh : 50

Num truck load of EVs between  Bengaluru  and  Trichy : 120

Num truck load of EVs between  Bengaluru  and  VIjayawada : 65

Num truck load of EVs between  Channai  and  Trichy : 305

Num truck load of EVs between  Delhi  and  Simla : 180

Num truck load of EVs between  Kolkata  and  Raipur : 95

Num truck load of EVs between  Kolkata  and  Agartala : 110

Num truck load of EVs between  Lucknow  and  Dibrugarh : 175

Num truck load of EVs between  Mumbai  and  Hassan : 205

Num truck load of EVs between  Mumbai  and  Simla : 90

Num truck load of EVs between  Mumbai  and  Raipur : 30

Num truck load of EVs between  Hyderabad  and  VIjayawada : 310

Num truck load of EVs between  Hyderabad  and  Raipur : 60

in this solution Lucknow-Simla transportation is 0 because cost is very high.

but after doing this total cost is increase
i.e Total cost : 2518200


# Exercise 1(10)

the Lucknow-Simla link, the Kolkata-Dibrugarh link and Delhi-Raipur
link is disrupted and no transportation is possible on the Lucknow-Simla link, the Kolkata-Dibrugarh link and Delhi-Raipur
link. Without changing the pyomo model, this is done by only change in .csv file, in csv file I will very incerese the Lucknow-Simla link, the Kolkata-Dibrugarh link and Delhi-Raipur
link transportation cost.
i.e new Lucknow-Simla transportation = 999999999, 
new Kolkata-Dibrugarh link = 999999999
new Delhi-Raipur link = 999999999

and new objecive function is : 

\begin{align}
\min  \ 827 x_{11} + 817 x_{12} + 1470 x_{13} + 2272 x_{14} + 1615 x_{15} + 3143 x_{16} + 1264 x_{17} + 4465 x_{18} + 559 x_{21} + 1832 x_{22} + 751 x_{23} + 3772 x_{24} + 637 x_{25} + 3773 x_{26} + 1856 x_{27} + 5459 x_{28} + 1109 x_{31} + 2443 x_{32} + 577 x_{33} + 3865 x_{34} + 798 x_{35} + 3521 x_{36} + 1984 x_{37} + 5303 x_{38} + 2943 x_{41} + 1471 x_{42} + 4676 x_{43} + 1337 x_{44} + 2702 x_{45} + 2741 x_{46} + 999999999 x_{47} + 4148 x_{48} + 3742 x_{51} + 3089 x_{52} + 2931 x_{53} + 3199 x_{54} + 2226 x_{55} + 999999999 x_{56} + 1691 x_{57} + 2128 x_{58} + 2353 x_{61} + 1160 x_{62} + 3295 x_{63} + 999999999 x_{64} + 2369 x_{65} + 2365 x_{66} + 1293 x_{67} + 3642 x_{68} + 1356 x_{71} + 1617 x_{72} + 2045 x_{73} + 2941 x_{74} + 2334 x_{75} + 4151 x_{76} + 1937 x_{77} + 4287 x_{78} + 195 x_{81} + 1862 x_{82} + 1331 x_{83} + 2921 x_{84} + 402 x_{85} + 2825 x_{86} + 563 x_{87} + 4441 x_{88} \nonumber \\
\end{align}

and constraints are same.

In [ ]:
data_csvfile.iloc[4,6]=999999999
data_csvfile.iloc[3,7]=999999999

In [ ]:
model_lab5_ex1.objective.deactivate()
model_lab5_ex1.objective = Objective(expr=sum(data_csvfile.iloc[i,j+1]*model_lab5_ex1.x[i,j] for i in row_indices for j in col_indices),sense=minimize)

This is usually indicative of a modelling error.
To avoid this warning, use block.del_component() and block.add_component().


In [ ]:
result = opt_cbc.solve(model_lab5_ex1)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [ ]:
# display solution
print('\nObjective = ', model_lab5_ex1.objective())

print('\nDecision Variables')
for i in row_indices:
  for j in col_indices:
    print('Num EVs between ',data_csvfile['Warehouse'][i],' and ', data_csvfile.columns[j+1],':', model_lab5_ex1.x[i,j].value)

print('\nConstraints')
model_lab5_ex1.constraints.display()


Objective =  2518200.0

Decision Variables
Num EVs between  Ahmedabad   and  Hassan : 0.0
Num EVs between  Ahmedabad   and  Jodhpur : 300.0
Num EVs between  Ahmedabad   and  Trichy : 0.0
Num EVs between  Ahmedabad   and  Simla : 0.0
Num EVs between  Ahmedabad   and  VIjayawada : 0.0
Num EVs between  Ahmedabad   and  Dibrugarh : 50.0
Num EVs between  Ahmedabad   and  Raipur : 0.0
Num EVs between  Ahmedabad   and  Agartala : 0.0
Num EVs between  Bengaluru  and  Hassan : 0.0
Num EVs between  Bengaluru  and  Jodhpur : 0.0
Num EVs between  Bengaluru  and  Trichy : 120.0
Num EVs between  Bengaluru  and  Simla : 0.0
Num EVs between  Bengaluru  and  VIjayawada : 65.0
Num EVs between  Bengaluru  and  Dibrugarh : 0.0
Num EVs between  Bengaluru  and  Raipur : 0.0
Num EVs between  Bengaluru  and  Agartala : 0.0
Num EVs between  Channai  and  Hassan : 0.0
Num EVs between  Channai  and  Jodhpur : 0.0
Num EVs between  Channai  and  Trichy : 305.0
Num EVs between  Channai  and  Simla : 0.0
Num EVs be

# Exercise 1(11)

Num EVs between  Ahmedabad   and  Jodhpur : 300

Num EVs between  Ahmedabad   and  Dibrugarh : 50

Num EVs between  Bengaluru  and  Trichy : 120

Num EVs between  Bengaluru  and  VIjayawada : 65

Num EVs between  Channai  and  Trichy : 305

Num EVs between  Delhi  and  Simla : 180

Num EVs between  Kolkata  and  Raipur : 95

Num EVs between  Kolkata  and  Agartala : 110

Num EVs between  Lucknow  and  Dibrugarh : 175

Num EVs between  Mumbai  and  Hassan : 205

Num EVs between  Mumbai  and  Simla : 90

Num EVs between  Mumbai  and  Raipur : 30

Num EVs between  Hyderabad  and  VIjayawada : 310

Num EVs between  Hyderabad  and  Raipur : 60

in this solution the Lucknow-Simla , the Kolkata-Dibrugarh  and Delhi-Raipur transportation is 0 because cost is very high.

but after doing this total cost is same as above solution because the Kolkata-Dibrugarh and Delhi-Raipur transportation have already zero means there is no truck to transport the EVs in this link.


i.e Total cost : 2518200
